substrate/runtime/session/src/lib.rs

```diff
@@ -135,15 +151,20 @@ impl<T: Trait> Module<T> {
 		// new set.
 		// check block number and call next_session if necessary.
 		let block_number = <system::Module<T>>::block_number();
-		if ((block_number - Self::last_length_change()) % Self::length()).is_zero() {
-			Self::rotate_session();
+		let is_final_block = ((block_number - Self::last_length_change()) % Self::length()).is_zero();
+		if is_final_block || Self::broken_validation() {
+			Self::rotate_session(is_final_block);
```

Comment:
I'm sure I missing something, but is this correct:
- `blocks_remaining()` returns zero if the session is due to rotate at the end of this block (c);
- `broken_validation()` returns `false` if `blocks_remaining()` is zero;
- so if we're calling `check_rotate_session()` at this 'final block of the session', noone is slashed, even though this final block was validated for eternity?

Reply:
`broken_validation` returns `true` iff we need to early-exit this session because too much time will have passed (calculated as a ratio > 1 of `session_length * block_period`), given that we would still need to wait at least `blocks_remaining * block_period`. `blocks_remaining` just returns the number of blocks, not including this that are left in the session, not counting the need for an early exit.

Reply:
but to answer the point, yes, we don't bother checking lateness if we're already exiting the session. we only consider it as being important if the session needs to exit early.

Reply:
in principle that could be changed...